### PR TITLE
Feature - Highlights Words

### PR DESCRIPTION
**Pull request by Johann-Snapp**
I've worked on #2190 and now create a pull request to get a review & integration of this feature.

I've created a package for the parsing associated to this feature.
I've worked in the callbacks sendNotificationsOnMessage & notifyUsersOnMessage in order to notify users when highlight words are found.
I've added a field in the admin panel so that the user can configure his highlight words.
Concerning the data model, highlight words are placed in `someUser.settings.preferences.highlights` as an array of strings.

### CODE REVIEW

**Comment by graywolf336** [comment]
One thing that you might want to change, to keep in sync with all the current packages, is change the package name from `rocketchat-highlight-words` to `rocketchat:highlight-words`.

File: .meteor/packages [hunk 1]
```diff
@@ -51,7 +51,6 @@ rocketchat:favico
 rocketchat:file
 rocketchat:github-enterprise
 rocketchat:gitlab
-rocketchat:highlight
```
Comment (sampaiodiego) [inline .meteor/packages]: please, add this package back. this package does code highlight which we don't want to remove :smile:

File: packages/rocketchat-lib/server/lib/notifyUsersOnMessage.js [hunk 2]
```diff
@@ -1,10 +1,36 @@
...
 		RocketChat.models.Subscriptions.incUnreadOfDirectForRoomIdExcludingUserId(message.rid, message.u._id, 1);
 	} else {
-		var mentionIds = [];
-		var toAll = false;
+		var mentionIds, toAll, highlightsIds, highlights;
+
+		mentionIds = [];
+		highlightsIds = [];
+		toAll = false;
+		highlights = RocketChat.models.Users.find({}, { fields: { '_id': 1, 'settings.preferences.highlights': 1 }}).fetch();
```
Comment (sampaiodiego) [inline packages/rocketchat-lib/server/lib/notifyUsersOnMessage.js]: we need to improve this query the most we can.. the main problems:
- this query will run on **every** message, so: 
  - we need to use a filter with an index.
  - we can't fetch() all users and loop all of them on every message sent
- we have to filter only users that can actually **see** the message (you don't want to notify someone not in the private group :smile: )

**Comment by engelgabriel** [comment]
@Johann-Snapp thanks for the valuable contribution. For the purpose of clarity, if it is ok with you, could you close this PR and open a new one with only the addition of the folder rocketchat:highlight-words?

**Comment by geekgonecrazy** [comment]
@Johann-Snapp thanks for this contribution :+1: this is awesome!

**Comment by Johann-Snapp** [comment]
Here it is: #2236

**Pull request closed without merging**